combinetf2/physicsmodels.py

```diff
@@ -0,0 +1,399 @@
+            for idx in sorted(self.proj_idxs, reverse=True):
+                cov = tf.reduce_sum(cov, axis=idx)
+                cov = tf.reduce_sum(cov, axis=idx + len(shape) - 1)
+            cov = tf.transpose(
+                cov, perm=perm_idxs + [i + len(perm_idxs) for i in perm_idxs]
+            )
+
+            if self.normalize:
+                norm = tf.reduce_sum(val)
+                jac = tf.eye(tf.shape(val)[0]) / norm - tf.expand_dims(val, 1) / norm**2
+                cov = tf.matmul(jac, tf.matmul(cov, tf.transpose(jac)))
+
+            var = tf.linalg.diag_part(cov)
+        else:
+            var = data[self.start : self.stop]
+
+            if self.normalize:
+                norm = tf.reduce_sum(var)
+                var = var / (norm**2)
+                # Additional variance from the normalization uncertainty
+                var = var + var**2
+
+            var = tf.reshape(var, shape)
+            var = tf.reduce_sum(var, axis=self.proj_idxs)
+            var = tf.transpose(var, perm=perm_idxs)
+
+        return val, var
+
+
+class Term:
+    def __init__(
+        self,
+        indata,
+        channel,
+        processes=[],
+        selections={},
+    ):
+        info = indata.channel_info[channel]
+
+        self.start = info["start"]
+        self.stop = info["stop"]
+
+        channel_axes = info["axes"]
+
+        channel_axes_names = [a.name for a in channel_axes]
+
+        self.has_data = not info["masked"] and len(processes) == 0
+
+        self.exp_shape = tuple([len(a) for a in channel_axes])
+
+        if processes is not None:
+            if any(p not in indata.procs.astype(str) for p in processes):
+                raise RuntimeError(
+                    f"Not all selection processes found in channel. Selection processes: {processes}, Channel axes: {indata.procs}"
+                )
+            self.proc_idxs = [
+                i for i, p in enumerate(indata.procs.astype(str)) if p in processes
+            ]
+        else:
+            self.proc_idxs = None
+
+        if selections is not None:
+            if any(k not in channel_axes_names for k in selections.keys()):
+                raise RuntimeError(
+                    f"Not all selection axis found in channel. Selection axes: {selections.keys()}, Channel axes: {channel_axes_names}"
+                )
+            self.selections = tuple(
+                [
+                    selections[n] if n in selections.keys() else slice(None)
+                    for i, n in enumerate(channel_axes_names)
+                ]
+            )
+            channel_axes = [c for c in channel_axes if c.name not in selections.keys()]
+            self.selection_idxs = [
+                i for i, n in enumerate(channel_axes_names) if n in selections.keys()
+            ]
+        else:
+            self.selections = None
+            self.selection_idxs = None
+
+        self.channel_axes = channel_axes
+
+
+class Ratio(PhysicsModel):
```

Comment:
This class doesn't provide a get_data method, is it optional in the API?  (and what does the code do if it's missing?)

Reply:
The get_data is inherited from PhysicsModel which calls self.make_fun, in this way the ratio for data should be correct in case it's done inclusively (w/o dividing processes where data makes no more sense). Of course the variance is not correct, so eventually a get_data would need to be implemented. But I'm thinking if the get_data can be derived from the make_fun also for the variance so I didn't do more work on this for now